source/Octopus.Tentacle/Kubernetes/KubernetesJobMonitor.cs

```diff
@@ -0,0 +1,165 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using k8s;
+using k8s.Models;
+using Octopus.Diagnostics;
+using Octopus.Tentacle.Contracts;
+
+namespace Octopus.Tentacle.Kubernetes
+{
+    public interface IKubernetesJobMonitor
+    {
+        Task StartAsync(CancellationToken token);
+    }
+
+    public interface IKubernetesJobStatusProvider
+    {
+        JobStatus? TryGetJobStatus(ScriptTicket scriptTicket);
+    }
+
+    public class KubernetesJobMonitor : IKubernetesJobMonitor, IKubernetesJobStatusProvider
+    {
+        readonly IKubernetesJobService jobService;
+        readonly IKubernetesPodService podService;
+        readonly ISystemLog log;
+        readonly Dictionary<ScriptTicket, JobStatus> jobStatusLookup = new();
+
+        public KubernetesJobMonitor(IKubernetesJobService jobService, IKubernetesPodService podService, ISystemLog log)
+        {
+            this.jobService = jobService;
+            this.podService = podService;
+            this.log = log;
+        }
+
+        async Task IKubernetesJobMonitor.StartAsync(CancellationToken cancellationToken)
+        {
+            while (!cancellationToken.IsCancellationRequested)
+            {
+                //initially load all the jobs and their status's
+                var initialResourceVersion = await InitialLoadAsync(cancellationToken);
+
+                //we start the watch from the resource version we initially loaded. This means we receive w
+                await jobService.WatchAllJobsAsync(initialResourceVersion,async (type, job) =>
```

Comment:
As mentioned, this is how to safely retrieve the status. You do an initial load and get the `resourceVersion` for when the list was loaded, you then start the `watch` from that point, so if there are any changes in-between the list being loaded and the watch starting, they will be immediately retrieved in the watch.